frontend/src/metabase/visualizer/utils/get-initial-state-for-card-data-source.ts

```diff
@@ -0,0 +1,70 @@
+import { isNotNull } from "metabase/lib/types";
+import { getColumnVizSettings } from "metabase/visualizations";
+import type { Card, DatasetColumn } from "metabase-types/api";
+import type { VisualizerHistoryItem } from "metabase-types/store/visualizer";
+
+import {
+  copyColumn,
+  createVisualizerColumnReference,
+  extractReferencedColumns,
+} from "./column";
+import { createDataSource } from "./data-source";
+
+export function getInitialStateForCardDataSource(
+  card: Card,
+  columns: DatasetColumn[],
+): VisualizerHistoryItem {
+  const state: VisualizerHistoryItem = {
+    display: card.display,
+    columns: [],
+    columnValuesMapping: {},
+    settings: {},
+  };
+  const dataSource = createDataSource("card", card.id, card.name);
+
+  columns.forEach(column => {
+    const columnRef = createVisualizerColumnReference(
+      dataSource,
+      column,
+      extractReferencedColumns(state.columnValuesMapping),
+    );
+    state.columns.push(copyColumn(columnRef.name, column));
+    state.columnValuesMapping[columnRef.name] = [columnRef];
+  });
+
+  const entries = getColumnVizSettings(card.display)
+    .map(setting => {
+      const originalValue = card.visualization_settings[setting];
+
+      if (!originalValue) {
+        return null;
+      }
+
+      if (Array.isArray(originalValue)) {
+        return [
+          setting,
+          originalValue.map(originalColumnName => {
+            const index = columns.findIndex(
+              col => col.name === originalColumnName,
+            );
+            return state.columns[index].name;
+          }),
+        ];
+      } else {
+        const index = columns.findIndex(col => col.name === originalValue);
+        if (!state.columns[index]) {
+          return;
```

Comment:
I believe it should be `return null` to be filtered properly.

Reply:
I don't think so, `isNotNull` uses a loose equality operator, therefore `isNotNull(undefined)` will return `false`.

Reply:
Ah, sorry, I searched the codebase when checking and got confused by `isNotNull` from `release/src/milestones.ts` which does strict equality. But it isn't even exported 😅